client/lib/docker.ts

```diff
@@ -0,0 +1,201 @@
+				Source:   reports,
+				Type:     "volume", 
+				ReadOnly: false
+			}
+		]
+
+		// Fill in the remaining options - we do that here as we need the port name for the env variables + volume name
+		coreOpts.Env = env;
+		coreOpts.HostConfig.Mounts = mounts;
+		workerOpts.Env = env;
+		workerOpts.HostConfig.Mounts = mounts;
+
+		await this.docker.createVolume(coreStorage);
+		await this.docker.createVolume(reports);
+		let core = await this.createContainer(`${__dirname}/../../../core`, `core`, corePort, coreOpts);
+		let worker = await this.createContainer(`${__dirname}/../../../worker`, `worker`, workerPort, workerOpts);
+
+		this.containerArray.push(
+			{
+				core: core,
+				worker: worker,
+				volumes:[
+					coreStorage,
+					reports
+				]
+			}
+		)
+	}
+
+
+	public async teardown(){
+		try{
+			for(let container of this.containerArray){
+				console.log(`Stopping ${container.core.id}...`)
+				await container.core.stop();
+				console.log(`Removing ${container.core.id}...`)
+				await container.core.remove();
+				console.log(`Stopping ${container.worker.id}...`)
+				await container.worker.stop();
+				console.log(`Removing ${container.worker.id}...`)
+				await container.worker.remove();
+	
+				// remove volumes
+				console.log(`Removing volumes`)
+				for (let volumeName of container.volumes){
+					let volume = await this.docker.getVolume(volumeName);
+					await volume.remove({force: true});
+				}
+			}
+		} catch(e){
+			console.log(e)
+		}
+	}
+
+	private async createContainer(dir: string, name: string, port: number, opts: any) {
+		console.log(`Creating ${name} container, listening on port ${port}`);
+		let imgTag = `${name}_${port}`;
+		let archive = `${name}.tar`;
+		
+		// we must create an archive of the core directory to be able to use it with dockerode build
+		if(!existsSync(archive)){
+			console.log(`Creating archive of container files...`);
+			let pack = tar.pack(dir).pipe(createWriteStream(archive));
+			await new Promise((resolve, _reject) => {
+				pack.on(`finish`, () => {
+					console.log(`Finished creating archive!`);
+					resolve();
+				})
+			});
+		} else{
+			console.log(`Archive of container files already exists - skipping...`)
+		}
+
+		console.log(`Building container....`)
+		// build container - after the first time it will build from cache
+		let stream = await this.docker.buildImage(
+			archive,
+			{t: imgTag, buildargs: {SKIP_INSTALL_BINARY: "true"}}
+		);
+
+		console.log(`Container ${imgTag} built! Starting container...`)
+		// this monitors the build progree, and waits until the build is finished before continuing
+		await new Promise((resolve, reject) => {
+			this.docker.modem.followProgress(stream, (err, res) => err ? reject(err) : resolve(res));
+		});
+
+		// create container, with ports env vars
+		opts.Image = imgTag;
+		const container = await this.docker.createContainer(
+			opts
+		);
+		await container.start()
+		console.log(`Container ${imgTag} started!`)
```

Comment:
This comment is deceptive as `${imgTag}` is not the name of the container, but the image tag the container is based on. Would be better to actually print the name of the container which I think is random because we are not setting `name` in our call to createContainer.